janitor/engineering.py

```diff
@@ -0,0 +1,135 @@
+
+
+@pf.register_dataframe_method
+def convert_units(
+    df: pd.DataFrame,
+    column_name: str = None,
+    existing_units: str = None,
+    to_units: str = None,
+    dest_column_name: str = None,
+) -> pd.DataFrame:
+    """
+    Converts a column of numeric values from one unit to another.
+
+    Functional usage example:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor.engineering
+
+        df = pd.DataFrame(...)
+
+        df = janitor.engineering.convert_units(
+            df=df,
+            column_name='temp_F',
+            existing_units='degF',
+            to_units='degC',
+            dest_column_name='temp_C'
+        )
+
+    Method chaining usage example:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor.engineering
+
+        df = pd.DataFrame(...)
+
+        df = df.convert_units(
+            column_name='temp_F',
+            existing_units='degF',
+            to_units='degC',
+            dest_column_name='temp_C'
+        )
+
+    Unit conversion can only take place if the existing_units and
+    to_units are of the same type (e.g., temperature or pressure).
+    The provided unit types can be any unit name or alternate name provided
+    in the unyt package's Listing of Units table:
+    https://unyt.readthedocs.io/en/stable/unit_listing.html#unit-listing.
+
+    Volume units are not provided natively in unyt.  However, exponents are
+    supported, and therefore some volume units can be converted.  For example,
+    a volume in cubic centimeters can be converted to cubic meters using
+    existing_units='cm**3' and to_units='m**3'.
+
+    This method mutates the original DataFrame.
+
+    :param df: A pandas dataframe.
+    :param column_name: Name of the column containing numeric
+        values that are to be converted from one set of units to another.
+    :param existing_units: The unit type to convert from.
+    :param to_units: The unit type to convert to.
+    :param dest_column_name: The name of the new column containing the
+        converted values that will be created.
+
+    :returns: A pandas DataFrame with a new column of unit-converted values.
+    """
+
+    # Check all inputs are correct data type
+    check("column_name", column_name, [str])
+    check("existing_units", existing_units, [str])
+    check("to_units", to_units, [str])
+    check("dest_column_name", dest_column_name, [str])
+
+    # Check that column_name is a numeric column
+    if not np.issubdtype(df[column_name].dtype, np.number):
+        raise TypeError(f"{column_name} must be a numeric column.")
+
+    # Check that existing_units and to_units are of the same type
+    # Need to handle area and volume units differently
+    if ("**" in existing_units) & ("**" in to_units):
+        # Check for same exponent on units
+        if existing_units.split("**")[1] == to_units.split("**")[1]:
+            existing_type = eval(
```

Comment:
I'm a bit wary of evaluating strings, even though I know this is a relatively restricted thing. 

Is there an alternative to evaluating the raw string? Could we maybe use a mapping (i.e. dictionary), e.g. like how it is done in janitor.chemistry?

(If my question doesn't make sense, please do point it out where I might be assuming something wrong!)

Reply:
I'm not entirely sure what you are thinking of in terms of code insecurities; I'd love to hear an example if you have one (mainly because I haven't really considered this issue in my coding before).

At any rate, that whole section of code in the method was meant to just provide more concise, better feedback in the case of a user trying to convert between incompatible units.  Of course, the unyt package will throw an error if the provided units are not both valid and compatible.  Therefore, I've removed that section of my code and just allowed the unyt error message to prevail (in this case, it will throw a UnitConversionError).

As a side effect, removing those lines of code dropped the test code coverage from 94% to 90%.  The only untested part of the submodule is a try-error construct in the case that the user does not have unyt installed.  I'm not sure if there's an easy way to have Pytest check for somthing like that (maybe a decorator that assumes a most basic python environment?).

Reply:
@rahosbach I think the main concern I have stems from injecting malicious code into an `eval` block. Given the code you’ve written, it’s hard for that to happen, but I think we can eliminate this attack surface by avoiding the use of `eval` in general. 

This SO comment has a great explanation why: https://stackoverflow.com/a/9384005

@anzelpwj has a nifty suggestion below to use `hasattr` and `getattr`. Perhaps we could use that instead, to safely access a module’s functions?

Reply:
Thanks @anzelpwj and @ericmjl.  I misinterpreted what your concern was at the start, but I can see why using eval() could be problematic (and thanks for the SO link).

The good news is that, since we are now just relying on unyt's error catching to throw UnitConversion errors whenever invalid or incompatible units are provided, we can just change the syntax that we use as follows:
**old**: `eval('unyt.' + existing_units)`
**new**: `unyt.Unit(existing_units)`

That was the only instance of eval() that still remained, and it's now gone.

Reply:
I will say that `NamedTuple` (Python core library even!) uses `eval`, so it definitely has a time and a place, but no need to run it if it's not needed.